plugins/newsdownloader.koplugin/internaldownloadbackend.lua

```diff
@@ -0,0 +1,56 @@
+local http = require("socket.http")
+local https = require("ssl.https")
+local logger = require("logger")
+local ltn12 = require("ltn12")
+local socket_url = require("socket.url")
+
+local InternalDownloadBackend = {}
+
+local function processDownloadAndReturnSink(url)
+    local request, sink = {}, {}
+    request['sink'] = ltn12.sink.table(sink)
+    request['url'] = url
+    local parsed = socket_url.parse(url)
+
+    local httpRequest = parsed.scheme == 'http' and http.request or https.request
+    -- first argument returned by skip is code
+    local _, headers, status = socket.skip(1, httpRequest(request))
+
+    -- raise error message when network is unavailable
+    if headers == nil then
+        error("Network is unreachable")
+    end
+
+    if status ~= "HTTP/1.1 200 OK" then
+        if status and string.sub(status, 1, 1) ~= "3" then -- handle 301, 302...
+           if headers and headers["location"] then
+            local redirected_url = headers["location"]
+            logger.dbg("HTTP response code: ", status)
+            logger.dbg("Redirecting to url: ", redirected_url)
+            -- clean up stuff, and recursively recall your download function
+            -- (you may want to add a counter, so you're not in an infinite loop
+            -- if the remote server is misconfigured and keeps redirecting)
+            return processDownloadAndReturnSink(redirected_url)
+           end
+        else
+           logger.warn("translator HTTP status not okay:", status)
+           return
+        end
+    end
+    return sink
+end
+
+function InternalDownloadBackend:getResponseAsString(url)
+    local sink = processDownloadAndReturnSink(url)
+    return table.concat(sink)
+end
+
+
+function InternalDownloadBackend:download(url, path)
+   local sink = processDownloadAndReturnSink(url)
+   ltn12.sink.file(io.open(path, 'w'))
```

Comment:
I'm not entirely sure what's going on here.

1. `sink` is for use as an argument to LuaSocket: http://w3.impa.br/~diego/software/luasocket/ltn12.html#sink.table
You can return the table that sink created for you, not the sink. You're doing that of course, but you're naming the table itself `sink`, which I find confusing at best. It's not a sink. It's just a results table for the sink.
2. Regardless of all that, the `ltn12.sink` here is unrelated to the `local sink` right above it, isn't it?
3. Anyway, so you'd probably want to `table.concat(table_name)` (please name it something not `sink`) just like you're doing in that getResponseAsString() function. But that probably makes more sense as `return table_name` on line 40.

Reply:
thanks I fixed it.